Task/allen/vtk9 master ospray

ARSanderson opened the pull request:
Initial pull request to review integration of VTK 9.1 and OSPRay 2.8 with the Kathleen's VTK 9.X.

This pull request only address code and not the needed bv support (bv_vtk and bv_ospray) those files have been separately given to Kathleen to integrate directly into her investigate_vtk9_master branch.

This code have been tested only on OS X. The goal is once integrated that further testing on linux clusters can be done.

ARSanderson committed:
initial changes for finding OSPRay 2.5

ARSanderson committed:
initial changes for OSPRay 2.5 and VTK 9.0 cmake

ARSanderson committed:
white space clean up and removal of unused include

ARSanderson committed:
replaced the polydata maker with a dtaset maker

ARSanderson committed:
update for including VTK 9.0

ARSanderson committed:
added a group box to better demarcate the OSPRay functionality

ARSanderson committed:
updates for OSX 11.2, VTK master and OSPRay 2.5

ARSanderson committed:
new class to parallel the vtkVisItDataSetMapper for OSPRay

ARSanderson committed:
added work-arounds for the OSPRay-vtCamera pan-zoom issue, and added a rotate to the Zoom mode

ARSanderson committed:
changed the perspective flaf to viewIs3D as the view could be perspective or orthographic, also using a local vtkOSPRayVisItDataSetMapper to over the vtkVisItDataSetMapper which removes a patch to vtk

ARSanderson committed:
inital VTK pipeline

ARSanderson committed:
fixed the transform issue

ARSanderson committed:
fixed the opacity issue

ARSanderson committed:
added the max limit

ARSanderson committed:
started more OSPRay

ARSanderson committed:
modifications for OSPRay and changed to nullptr

ARSanderson committed:
modifications for OSPRay

ARSanderson committed:
modifications for OSPRay

ARSanderson committed:
removed unused var and mods for OSPRay resampling

ARSanderson committed:
Added OSPRay to be optional

ARSanderson committed:
additional rework for OSPRay

ARSanderson committed:
minor cleanup

ARSanderson committed:
added check for number of components needed

ARSanderson committed:
added different resampling options

ARSanderson committed:
added a per rank sampling option

ARSanderson committed:
added a per rank sampling option

ARSanderson committed:
changed from view to viewInfo for clarity

ARSanderson committed:
first true version of a working VTK - OSPRay renderer

ARSanderson committed:
clean as OSPRay is now more integrated

ARSanderson committed:
removed OSPRay 1 rendering

ARSanderson committed:
renamed files to remove OSPRay which is part of VTK

ARSanderson committed:
clean up of the class to reduce code duplication

ARSanderson committed:
renamed member variabels for clarity

ARSanderson committed:
minor cleanup

ARSanderson committed:
more accurate name

ARSanderson committed:
minor whitespace clean up

ARSanderson committed:
more accurate name

ARSanderson committed:
removed debug cerr

ARSanderson committed:
added default values

ARSanderson committed:
add ray tracing

ARSanderson committed:
added default values and changed to int for consistancy

ARSanderson committed:
added all ospray rendering attribute gets

ARSanderson committed:
more accurate name

ARSanderson committed:
added depth peeling and ospray rendeing attributes

ARSanderson committed:
added default values

ARSanderson committed:
added ospray rendeing attributes and reorganize to assure all rendering attributes were accounted for

ARSanderson committed:
added depth peeling and ospray rendering attributes

ARSanderson committed:
vtk ray tracing is not dependent on ospray

ARSanderson committed:
white space adjustment

ARSanderson committed:
added get data tree to be public

ARSanderson committed:
whitespace clean up and added a missed initial value

ARSanderson committed:
missed class scope

biagas commented on src/avt/Filters/CMakeLists.txt:
```diff
@@ -176,33 +170,23 @@ ${VTKh_INCLUDE_DIRS}
-        TARGET_LINK_LIBRARIES(avtfilters_ser ${ICET_FULLPATH_LIB})
-    ENDIF()
-ENDIF() 
+TARGET_LINK_LIBRARIES(avtfilters_ser visitcommon lightweight_visit_vtk visit_vtk avtmath avtview avtpipeline_ser avtdatabase_ser avtdbatts VTK::CommonMath VTK::CommonTransforms VTK::FiltersGeneral VTK::FiltersModeling VTK::IOXML VTK::RenderingRayTracing VTK::RenderingVolumeOpenGL2)
```
I am concerned about introducing a direct OpenGL dependency to avtFilters.
Could the relevant classes depending on OpenGL live in avt/Plotter instead?

The separation of direct dependencies with OpenGL is important now with the need for separate Mesa and OSMesa versions of VTK for our HPC systems.

ARSanderson replied:
Good question, I have no idea. I am not sure why these Volume rendering classes are in avt/Filters in the first place.

Commit:
Merge branch 'task/biagas/investigate_vtk9_master' into task/allen/vtk9_master_ospray

biagas commented on src/avt/Filters/avtICETCompositor.h:
```diff
@@ -0,0 +1,54 @@
+// details.  No copyright assignment is required to contribute to VisIt.
+
+// ************************************************************************* //
+//                           avtVisItVTKCompositor.h                         //
```
Update to reflect actual name.

ARSanderson replied:
Done

biagas commented on src/avt/Filters/avtICETCompositor.h:
```diff
@@ -0,0 +1,54 @@
+#endif
+
+class avtICETCompositor
+{
```
Need exports header and API specification in class declaration.
If this class remains here in filters (see my comment in CMakeLists.txt) , 
then include `filters_exports.h` and use
`class AVTFILTERS_API avtICETCompositor` for class declaration

ARSanderson replied:
Done

biagas commented on src/avt/Filters/avtICETCompositor.C:
```diff
@@ -0,0 +1,359 @@
+                                        IceTImage result)
+{
+    // int width = icetImageGetWidth(result);
+    // int height = icetImageGetHeight(result);
```
Are these commented out for future use? Or was it determined they aren't needed? If not needed, please remove.  Here and below in the else clause.

ARSanderson replied:
Will need to consult with Kevin ... initial code

biagas commented on src/avt/Filters/avtICETCompositor.C:
```diff
@@ -0,0 +1,359 @@
+    // const IceTFloat bgColor[4] =
+    // {
+    //     IceTFloat(0.0f), IceTFloat(0.0f), IceTFloat(0.0f), IceTFloat(0.0f)
+    // };
```
Please remove the commented out section.

ARSanderson replied:
Done

biagas commented on src/avt/Filters/avtICETCompositor.C:
```diff
@@ -0,0 +1,359 @@
+    s_width = width;
+    s_height = height;
+
+    debug5 << "[VisItVTK::Compositor] width = " << width << " height = " << height << std::endl;
```
Shouldn't this be avtIceTCompositor instead of VisItVTK::Compositor ? Here and below.

ARSanderson replied:
Fixed - missed in the renaming

biagas commented on src/avt/Filters/avtICETCompositor.C:
```diff
@@ -0,0 +1,359 @@
+    if(colorBuffer != nullptr)
+    {
+        memcpy(outBuffer, colorBuffer, bytesToCopy);
+        // avtVisItVTKRenderer::WriteArrayToPPM("test", renderedFrameBuffer, screen[0], screen[1]);
```
Should this be removed? If left in for future debugging help, perhaps add a comment to that effect and ensure it works if un-commented.

ARSanderson replied:
debug code

biagas commented on src/avt/Filters/avtResampleFilter.C:
```diff
@@ -561,15 +570,9 @@ avtResampleFilter::ResampleInput(void)
                     else
-                    {
-                        rg->GetPointData()->AddArray(vars[i]);
                         rg->GetPointData()->SetScalars(vars[i]);
```
This may require more investigation, but I believe SetScalars removes any array from the dataset that was previously designated as the Active scalars.  If removing arrays is a problem in this situation, then should use AddArray followed by SetActiveScalars(varname).  If removing arrays at this point isn't a problem, this comment can be ignored.

ARSanderson replied:
The original code had both an AddArray and a SetScalar. The AddArray is redundant as SetScalars does that.

int vtkDataSetAttributes::SetAttribute(vtkAbstractArray* aa, int attributeType)
// This method lets the user add an array and make it the current
// scalars, vectors etc... (this is determined by the attribute type
// which is an enum defined vtkDataSetAttributes)

That said, in this step arrays are being added to a new vtkRecGrid. If you look at the rest of the vector sniper SetVectors and SetTensors are used without the AddArray.

biagas replied:
The VTK documentation for the [SetAttribute method](https://vtk.org/doc/nightly/html/classvtkDataSetAttributes.html#afac42658dd14584d312f132556f83a96) specifically states that any array previously set for that attribute type will be removed from the DataSetAttributes.

The for-loop indicates it could possibly be adding multiple scalars/vectors/tensors. In which case, only the last added Scalar array will still be in the dataset attributes at then end of the for loop.  Am I misunderstanding the documentation? 

The original logic was probably wrong as well. It should be 'AddArray', followed by SetActiveScalars.
Actually, not sure the SetActiveScalars/Tensors/Vectors is even needed. Most of our filters operate on named arrays and only set the active attributes as needed.

biagas commented on src/avt/Filters/avtResampleFilter.C:
```diff
@@ -795,7 +798,12 @@ avtResampleFilter::GetDimensions(int &width, int &height, int &depth,
+    if (atts.GetPerRankResample())
+    {
+      avtDataset_p ds = GetTypedInput();
+      avtDatasetExaminer::GetSpatialExtents( ds, bounds );
```
Needs more spaces in indent to match else clause.

ARSanderson replied:
Done

ARSanderson committed:
fixed spelling mistake

ARSanderson committed:
changed to new VTK cmake format

ARSanderson committed:
removed need to check for ospray

ARSanderson committed:
fixed python atts

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+                // when switching to scene light from another light type.
+                // Can be removed when we switch to new VTK dated after
+                // 6Sep02.
+                //
```
Comment says the line shouldn't be needed with newer VTK. Did you investigate if this was true (looks like this was copied from VisWinLighting.

ARSanderson replied:
My initial albeit limited testing shows no change in the rendering so I have removed it.

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+        avtImageRepresentation::NewImage(width, height, nColorChannels);
+
+    finalImage->GetImage().SetImageVTK(finalImageData);
+    // finalImage->GetImage() = imageData;
```
Should this commented out code be removed?

ARSanderson replied:
Removed

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+    if(m_dataType == DataType::VOLUME)
+        ExecuteVolume();
+    else
+        ; // TODO: Implement ExecuteSurface(); // Done as PseudoColotPlot
```
spelling: PseudocolorPlot ?

ARSanderson replied:
Fixed

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+        m_nComponents = 1;
+    }
+
+    std::cerr << __LINE__ << " [VisItVTKDevice] "
```
Convert this to write to debug logs instead of stderr.  Here and all occurrences below as well.

ARSanderson replied:
This debugging code will be removed or perhaps converted.

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+        // There could be both a scalar and opacity data arrays.
+        vtkDataArray *dataArr = in_ds->GetPointData()->GetScalars();
+        vtkDataArray *opacityArr = nullptr;
+        // vtkDataArray *gradientArr = nullptr;
```
Is the commented out var (and logic which uses it below) set aside for future work?
If so, please comment, otherwise remove.

ARSanderson replied:
Future usage???

ARSanderson committed:
removed old comment

biagas commented on src/avt/Filters/avtVisItVTKDevice.C:
```diff
@@ -0,0 +1,1151 @@
+
+        {
+            vtkXMLImageDataWriter* writer = vtkXMLImageDataWriter::New();
+
```
Is this section for debugging? If so, perhaps wrap in an #ifdef ?

ARSanderson replied:
commented out

biagas commented on src/avt/Plotter/vtk/vtkOSPRayVisItDataSetMapperNode.C:
```diff
@@ -0,0 +1,109 @@
+  See Copyright.txt or http://www.kitware.com/Copyright.htm for details.
+
+     This software is distributed WITHOUT ANY WARRANTY; without even
+     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
```
Use VisIt's copyright lines.

ARSanderson replied:
How do we handle code that has come from VTK but we have modified???

biagas commented on src/avt/Plotter/vtk/vtkOSPRayVisItDataSetMapperNode.h:
```diff
@@ -0,0 +1,53 @@
+  Module:    vtkOSPRayVisItDataSetMapperNode.h
+
+  Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
+  All rights reserved.
```
Use VisIt's copyright lines.

ARSanderson replied:
How do we handle code that has come from VTK but we have modified???

biagas replied:
This actually is our code, subclassing one of theirs, so our copyright lines should be fine.

biagas commented on src/avt/VisWindow/Colleagues/VisWinRendering.h:
```diff
@@ -410,9 +412,9 @@ class VISWINDOW_API VisWinRendering : public VisWinColleague
-    vtkRenderer                  *foreground;
+    vtkRenderer                  *canvas {nullptr};
+    vtkRenderer                  *background {nullptr};
+    vtkRenderer                  *foreground {nullptr};
```
Why the need for initialization here? They are initialized in the constructor.

ARSanderson replied:
Fell swoop of initializing everything in the header regardless of what happens in the C file. Selective inits in the header are confusing so do them all.

biagas commented on src/avt/VisWindow/VisWindow/VisWindow.h:
```diff
@@ -810,21 +810,21 @@ class VISWINDOW_API VisWindow
-    VisCallback                       *hideMenuCallback;
-    void                              *hideMenuCallbackData;
+    VisCallback                       *showMenuCallback {nullptr};
+    void                              *showMenuCallbackData {nullptr};
```
why the initializers here when they are initialized in the constructor?

ARSanderson replied:
Fell swoop of initializing everything in the header regardless of what happens in the C file. Selective inits in the header are confusing so do them all.

biagas commented on src/avt/Pipeline/Data/avtDataset.h:
```diff
@@ -162,12 +162,14 @@ class PIPELINE_API avtDataset : public avtDataObject
 
     std::string              GetDatasetAsString();
+
+    avtDataTree_p           &GetDataTree() { return dataTree; };
```
Why was this necessary? Would adding a friend class to this list above have worked instead?

ARSanderson replied:
Not clear to me that it makes sense to make the parent class a friend class.

biagas commented on src/plots/Volume/QvisVolumePlotWindow.C:
```diff
@@ -883,45 +883,56 @@ QvisVolumePlotWindow::CreateOpacityGroup(QWidget *parent, QVBoxLayout *pLayout,
-        resampleTargetWidget            = new QWidget();
-        QHBoxLayout *resampleTargetLayout= new QHBoxLayout(                                 resampleTargetWidget);
+        resampleGroup->setTitle(tr("Resampling Options"));
+        // resampleGroup->setFlat(true);
```
If the commented out settings aren't needed, they should be removed.

ARSanderson replied:
Done

biagas commented on src/plots/Volume/QvisVolumePlotWindow.C:
```diff
@@ -1382,19 +1415,60 @@ void QvisVolumePlotWindow::CreateOSPRayGroups(QWidget *parent, QLayout *pLayout)
+    osprayLayout->addWidget( osprayToggle );
+    osprayLayout->addWidget( osprayProperties );
+
+    // osprayPropertiesLayout->addWidget(osprayShadowToggle,             0, 0);
```
Are these commented-out settings reserved for future use? If so please add a comment about under what circumstances they should be enabled, otherwise please delete them.

ARSanderson replied:
Not everything is enabled in OSPRay via VTK - comment added.

biagas commented on src/plots/Volume/avtDefaultRenderer.h:
```diff
@@ -46,17 +48,21 @@ class avtDefaultRenderer : public avtVolumeRendererImplementation
-    vtkVolumeProperty         *volumeProp;
-    vtkSmartVolumeMapper      *mapper;
+    vtkColorTransferFunction  *transFunc    {nullptr};
+    vtkPiecewiseFunction      *opacity      {nullptr};
```
Why initializers here, when they are initialized in the constructor?

ARSanderson replied:
Fell swoop of initializing everything in the header regardless of what happens in the C file. Selective inits in the header are confusing so do them all.

biagas commented on src/plots/Volume/avtDefaultRenderer.C:
```diff
@@ -144,128 +154,157 @@ void
-        double spacingZ = rgrid->GetZCoordinates()->GetTuple1(1)-
-            rgrid->GetZCoordinates()->GetTuple1(0);
+        // Check for an implied transform.
+        // avtDataAttributes &inatts = GetInput()->GetInfo().GetAttributes();
```
If the commented out section is reserved for future use, please add comments. Otherwise remove.

ARSanderson replied:
Added comment - this class will be deprecated at some point.

biagas commented on src/viewer/core/ViewerWindow.h:
```diff
@@ -802,26 +802,26 @@ class VIEWERCORE_API ViewerWindow : public ViewerBase
-    ViewerActionManager *actionMgr;
+    VisWindow           *visWindow {nullptr};
+    ViewerPlotList      *plotList {nullptr};
+    ViewerActionManager *actionMgr {nullptr};
```
Why initializers in header when they are initialized in constructor?

ARSanderson replied:
Fell swoop of initializing everything in the header regardless of what happens in the C file. Selective inits in the header are confusing so do them all.

ARSanderson committed:
removed the need for the VISIT_OSPRAY replaced with HAVE_OSPRAY

ARSanderson committed:
complete rewrite mininal checks needed now

biagas reviewed (changes requested):
I'm curious about all the places you added initializers to header files when there are initializers in the constructors.  I know this is a C++11 thing, could you explain the benefits? (Forgive me, I made comments in each header where you did this).  Is this something fully supported by the various compilers that will be used to build VisIt?

ARSanderson committed:
removed the need for the VISIT_OSPRAY

biagas commented:
@markcmiller86 @brugger1 @cyrush adding all of you as reviewers. Please feel free to remove yourself if you don't have time.
It's a pretty big change so want other eyes on it all.

ARSanderson committed:
removed the need for the VISIT_OSPRAY replaced with HAVE_OSPRAY

ARSanderson committed:
added visit-config.h wher needed and minor cleanup

ARSanderson commented:
> I'm curious about all the places you added initializers to header files when there are initializers in the constructors. I know this is a C++11 thing, could you explain the benefits? (Forgive me, I made comments in each header where you did this). Is this something fully supported by the various compilers that will be used to build VisIt?

By habit I have stared doing this everywhere. The benefit is that one can easily do the init where the var is defined. I was getting a crash but was not sure why. I looked at a a couple of header files,and added the init. Crash was gone. The alternative is to search through the C code for each var for where it first used. Just a better paradigm.

Commit:
Merge branch 'task/biagas/investigate_vtk9_master' into task/allen/vtk9_master_ospray

markcmiller86 commented:
> It's a pretty big change so want other eyes on it all.

Agreed. I don't want to contribute more to @ARSanderson work. OTOH, if anyone sees a nice way to [break this PR up into smaller pieces](https://bssw.io/items/pull-request-size-matters), that might be worth some time considering. I won't have a chance to take a look until middle next week.

ARSanderson commented:
> Agreed. I don't want to contribute more to @ARSanderson work. OTOH, if anyone sees a nice way to [break this PR up into smaller pieces](https://bssw.io/items/pull-request-size-matters), that might be worth some time considering. I won't have a chance to take a look until middle next week.

There are a few minor chunks but so small they are noise (avtResampleFilter). Perhaps the best way would be to do a group review going through the changes of significance. Kathleen has picked up on some areas.

ARSanderson committed:
cleanup of name and added export dec

ARSanderson committed:
white space cleanup

ARSanderson committed:
added comment to commented out code

ARSanderson committed:
removed old code and anded comment about future code

ARSanderson committed:
Added comment and control only debug vs cerr

ARSanderson committed:
switch to 9.1

ARSanderson committed:
made methods pure virtual so all are defined, fixed override warning and white space cleanup

ARSanderson committed:
fixed vtk 9.1 deprecations

ARSanderson committed:
added visit-config.h

ARSanderson committed:
cleaned up the two ray casting methods so they are similar, and removed unneeded gradient calc

ARSanderson committed:
name change to indicate member and OSPRay vars

ARSanderson committed:
pendatic white space include ordering clean up

ARSanderson committed:
switch to OSPRAY 2.8

ARSanderson committed:
removed old condition

ARSanderson committed:
removed conditional as we always use the the vtk ray tracing

ARSanderson committed:
added header file point inits

ARSanderson committed:
changed to nullptr

ARSanderson committed:
made the device by deleted with the class and changed to nullptr

ARSanderson committed:
added the ability to use the scivis and pathtracer

ARSanderson committed:
cleanup of the ability to regenerate on demand

ARSanderson committed:
moved to silo 4.11

ARSanderson committed:
fixed checks for the regeneration

ARSanderson committed:
added check for OSPRay support and better messaging

ARSanderson committed:
added check for OSPRay support

ARSanderson committed:
added OSPRay thirdparty libs

biagas committed:
Changes for Windows/Linux.

bv_ospray: add Allens changes
bv_vtk: uncomment URL, add osmesa dependency, add Allens ospray patches,
CMakeLists.txt, find OSPRay and Qt before VTK, remove old EMBREE, ISPC and TBB vars.
Update windows config-site files for ospray
avtVisItVTKDevice.C, add include for avtCallback header.

ARSanderson committed:
check the dir as the flag is dead

ARSanderson committed:
fixed case

ARSanderson committed:
added warning for the user when resampling

ARSanderson committed:
Change OSPRay group to use a flag, cleanup of the nullptr inits, and added warning

ARSanderson committed:
added warning

ARSanderson committed:
clean up of the constructor inits

ARSanderson committed:
fixed typo

ARSanderson committed:
comment update

ARSanderson committed:
added check for null scalar data

biagas committed:
Fix problem with HAVE_OSMESA not being defined.
Fix a few issues with Find VTK.
Comment out factory override for vtkOSOpenGLRenderWindow, not needed in current build scenario.

ARSanderson committed:
initial work to handle cell centered data

ARSanderson committed:
split ospray patching up as the second half of the patch was applied twice

ARSanderson committed:
added initial values

ARSanderson committed:
added initial values

ARSanderson committed:
cleaned up inital values, fined tune the opacity and lighting check

ARSanderson committed:
cleaned up when recalculation requirement

ARSanderson committed:
added local debug option and fined tune the lighting check

ARSanderson committed:
flipped to use debug5

ARSanderson committed:
flipped to use debug5

ARSanderson committed:
added support for color/opatiy min/max, added auto resample msg, and initial fix for cell data

ARSanderson committed:
added support for color/opatiy min/max, added auto resample msg

ARSanderson committed:
needed to use float

ARSanderson committed:
added auto resample flag

ARSanderson committed:
made the resamplig parallel the resample filter

ARSanderson committed:
reworked the resamplig logic

ARSanderson committed:
minor clean up of the centering

ARSanderson committed:
added legact code for atts

biagas committed:
Fix compile errors when VTK not built with OSPray support.
Make linking with VTK::RenderingRayTracing conditional upon its existence.

biagas committed:
Add vtk patch so that xmlversion header gets installed.

ARSanderson committed:
added the centering

ARSanderson committed:
Major surgery to remove unused gradient code and to conlapse the serial volume rendering

ARSanderson committed:
Removed RLE files

ARSanderson committed:
merge

ARSanderson committed:
added comments and re-organized the code

biagas committed:
Fix volume plots vtk_module_autoinit.

ARSanderson committed:
added more to the resampling test and user warnings

ARSanderson committed:
added more to the recentering/resampling test and user warnings

ARSanderson committed:
added HasData check, changed to nullptr, and whitespace cleanup

ARSanderson committed:
added a check for per rank resampling not to return anything when it had no data originally

ARSanderson committed:
cleanup of lighting and fixed when resampling was needed

ARSanderson committed:
minor comment and whitespace cleanup

markcmiller86 commented:
Remarks on current interface labels...

* For `cell` vs. `point` vs. `maintain` (although other parts of VisIt are not fully consistent on this - **Pseudocolor** plot uses `nodal` and `zonal` and `original` whereas **Label** plot uses `cell` and `node`, **Curve** uses `point`, in several places we have *clean `zones` only*), I would recommend `nodal` vs `zonal` vs. `native` (or `original`)
* For rendering options, I don't like `serial` and `parallel` or the repeated use of `Rendering` in either of those options. I dunno what the *salient* difference is in these options but I think anyone seeing them would always go for a *parallel* option. Why use *serial* ever? I think the answer is that it can be fast but maybe not accurate. I might call these `preview` and `accurate` instead. Or `coarse` and `fine` or something along those lines.

ARSanderson commented:
@markcmiller86 - I have gone with native, nodal, and zonal.

Regarding the rendering modes. One of the issues is when someone is running on a single rank (serial) the two rendering modes will produce the same quality of image. So coarse is fine, and fine is coarse. Accurate is not a good description either because the rendering is accurate regardless of the scale.

I think a better approach may be along the lines of Eric's comment about local vs scalable. But I am not sure.

ARSanderson committed:
moved rendering to the volume plot to remove open gl dependencies

ARSanderson committed:
moved rendering to the volume plot to remove open gl dependencies

ARSanderson committed:
changed to nodal, zonal, and native

ARSanderson committed:
switch to debug output

markcmiller86 commented:
> @markcmiller86 - I have gone with native, nodal, and zonal.
> 
> Regarding the rendering modes. One of the issues is when someone is running on a single rank (serial) the two rendering modes will produce the same quality of image. So coarse is fine, and fine is coarse. 

How is what happens in the single ranke parallel case relevant? First, who (users) runs VisIt with one processor parallel? No body except developers because there is some handiness in being able to do that. Pitching the interface so as to avoid a possible muddying of the diffs between these two modes in this one specific use case seems unreasonable. For me, its not relevant.

> Accurate is not a good description either because the rendering is accurate regardless of the scale.

Really? How can that be true when number of samples represents a tiny fraction of total dataset size? The input would be highly undersampled. I agree the code would do an *accurate* job of rendering those samples but the final output would be a poor approximation of the input due to sampling. But, its useful because it can be done *quickly*. To me, its like the difference between seeing just the bounding box (renders very fast) and seeing the scalable rendered image when VisIt is in SR mode...the bounding box is useful because its interactive/fast and can be used as a proxy for the actual object as far as view orientation, etc. is concerned.

> 
> I think a better approach may be along the lines of Eric's comment about local vs scalable. But I am not sure.

`Local` vs. `scalable` has the same problems `serial` vs. `parallel` does...why would any user opt to use `local` when `scalable` is an available option?

markcmiller86 commented:
I might argue that what we currently label here as `Serial Rendereing` is kinda sorta the **Volume** plot's equivalent of the bounding box for SR mode.

In fact, why not have the **Volume** plot just kinda sorta behave the same way...it uses the `Serial Rendering` stuff when user is adjusting view parameters, changing plot parameters which are orthogonal to **Volume** rendering details and when they release the mouse (or just wait long enough) it renders the *real* image using `Parallel Rendering` mode. In this case, the user doesn't explicitly pick one or the other...VisIt automatically chooses based on maintaininging interactivity for key paramater setting steps and automatically renders a full res image when user's parameter changes settle out.

ARSanderson commented:
I like the idea of the "serial" mode being used as an interactive preview. There would definitely need to be some user flexibility so to get interactivity. 

Resolution issues aside the serial version does not produce exactly the same image as the other renders (parallel, compositing, SLVR) due the composting and other algorithmic differences, so users might object to the differences. One would expect a "pop" in image resolution but not the color. Kevin and I need to further discuss the compositing as he may have figure out a a way to limit some of this issue so that the parallel and serial are closer together.  But the "composting" and SLVR generate very different images. 

Regarding my comment on "accurate" I was not clear but you parsed through it. The rendered image is accurate based on the immediate data the rendering is given. That immediate data could be the native input data, up sampled or down sampled from the input data. So what does accurate mean???

markcmiller86 commented:
> I like the idea of the "serial" mode being used as an interactive preview. There would definitely need to be some user flexibility so to get interactivity.

When we first developed *Scalable Rendering* (SR) mode, we had to decide on a *polygon count* threshold above which, VisIt would automagically transition into SR mode and below which, VisIt would use *normal* (local GL hardware) rendering. We have a 3-way control for that (`Always`, `Auto`, `Never`) where SR mode is always on, regardless of polygon count or it is never on or VisIt decides (`Auto`) when to turn it on.

Would something similar apply here? In `Always`, user would always get what you now call `Parallel Rendering`. In `never`, user would always get what you now call `Serial Rendering` and in `Auto` VisIt would automatically choose which based on size of input (maybe the ratio of input #nodes to #samples being used in the *serial* rendering). If so, we might wanna just call this `Scalable VR` with the Always/Auto/Never choice and then the other options currently in the rendering pull-down don't have the Always/Auto/Never feature.

Honestly, that sounds like it could be more work than its worth, especially getting the viewer and engine to handle auto-matic transitions into and out of `Scalable VR` (thats a particularly challenging bit of software engineering for SR mode) and just picking clearer names for the two modes would be fine as part of the initial implementation.

biagas committed:
Remove the API designation.

ARSanderson committed:
reduced when to regenerate

ARSanderson committed:
reduced when to regenerate

ARSanderson committed:
change to do all resampling in the plot. Made the two renders as similar as possible for future collapsing

ARSanderson committed:
Merge branch 'task/allen/vtk9_master_ospray' of github.com:visit-dav/visit into task/allen/vtk9_master_ospray

ARSanderson committed:
now using the VolumeAtts directing in the rendering and fixed the OSPRay capitalization

ARSanderson committed:
removed the factory as it is not needed

ARSanderson committed:
refactored the rendering into a bass class

ARSanderson committed:
minor revisions for teh combined class refactoring.

ARSanderson committed:
forgot to remove the debug code

ARSanderson committed:
attempt to deal with upstream sampling

ARSanderson committed:
updated the volume plot page

ARSanderson committed:
added new volume rendering tests

ARSanderson committed:
removed local debug

ARSanderson committed:
added a check for two component rendering

markcmiller86 commented:
@visit-dav/visit-developers might we want to divide and conquer here? If anyone has suggestions on best division, please suggest one. Theres 120 files and about 20,000 lines changed.

biagas commented:
@markcmiller86 not sure the best way to divide.  Good tip regardless of what you are examining is to use the 'turn off whitespace' filter, that will help. I am going to test out the branch on Windows before I review.

markcmiller86 commented:
We have some conflicts preventing this branch from being up to date. I haven't looked at them here but I think I can handle the conflicts in the 2 `.C` files. Can you take a look at the `.cmake` and `bv_vtk.sh` file conflicts? Is it even possible to split up this work via the web interface?

markcmiller86 commented:
Ok, I just tried the `Resolve conflicts` button and yes, I can select which files I work on there. So, it is possible to split this activity up.

markcmiller86 commented:
> Ok, I just tried the `Resolve conflicts` button and yes, I can select which files I work on there. So, it is possible to split this activity up.

Umm...Nope! I am wrong 😢 . You can't commit until you've resolved all conflicts.

And, in looking closely, I see that I volunteered for resolving just a few 1-line conflicts leaving you the bulk of the work so it doesn't even help ease the load relative to the others.

biagas commented:
I will resolve the conflicts via my checkout of the branch and push the changes.

biagas committed:
Squashed commit of the following:

commit 96ef4fd07674d5e6aa0c83ef5bde8f87423e6ece
Author: biagas <biagas2@llnl.gov>
Date:   Mon Jan 31 11:15:50 2022 -0800

    Fix patch for vtkCutter.

commit 4bd7bda574ea9da51836046bdf075e5d0c63ea7e
Author: biagas <biagas2@llnl.gov>
Date:   Mon Jan 31 09:23:17 2022 -0800

    Apply a patch to vtkCutter.
    Prevents problems with the cutter changing the datatypes of cell and point data arrays.

commit a8b438b5bf637338d8f54a005656a497473964cb
Author: biagas <biagas2@llnl.gov>
Date:   Thu Jan 27 12:35:56 2022 -0800

    Add patch for vtkRectilinearGridReader.cxx,
    prevents unnecessary vtk warning.

commit 6a68826805ad242232e40961eb4acdb2758f603f
Author: biagas <biagas2@llnl.gov>
Date:   Tue Jan 25 13:46:26 2022 -0800

    Fix typo.

commit b839484febd1dbc2e8ed474a9204cd22eb7a6ebb
Author: biagas <biagas2@llnl.gov>
Date:   Tue Jan 25 08:12:57 2022 -0800

    Add RequiredDatabasePlugin call for OpenFOAM test.

commit 153e025fa6cd6036d75e01860728939ae079aa44
Author: biagas <biagas2@llnl.gov>
Date:   Tue Jan 25 07:54:28 2022 -0800

    Update mfem version to 4.3

commit 4712f1d26184415773d5398179c9285b1c918e7d
Author: biagas <biagas2@llnl.gov>
Date:   Mon Jan 24 20:08:50 2022 -0800

    Fix warnings related to vtk API changes.

commit ba762fb5dc673aa773632c7aa497381d8a931abe
Author: biagas <biagas2@llnl.gov>
Date:   Mon Jan 24 11:35:41 2022 -0800

    Fix patch for installation of vtk's xmlversion header.

commit 233ed4ed07723056a03746e8b33f615c4992ed6b
Author: biagas <biagas2@llnl.gov>
Date:   Wed Jan 12 11:33:57 2022 -0800

    Add patch for installation of xml_version.h

    So that Xdmf can build successfully (it uses VTK's xml).

commit 14791fd2eaae7e6a36780068e5820b03c98a75d1
Author: biagas <biagas2@llnl.gov>
Date:   Wed Jan 12 10:43:58 2022 -0800

    Minor tweaks for VTK/QT find and build.

ARSanderson committed:
reverted number as test is generated

ARSanderson committed:
Merge branch 'task/allen/vtk9_master_ospray' of github.com:visit-dav/visit into task/allen/vtk9_master_ospray

ARSanderson committed:
added initial baseline images, note 10 and 20 are blank due to an unknonw bug in pipeline

biagas commented on src/common/state/RenderingAttributes.h:
```diff
@@ -9,7 +9,7 @@
 #include <ColorAttribute.h>
-#include <float.h>
+#include <visit-config.h>
 
```
Why was this needed in the header?

ARSanderson replied:
It was added there because many other files include it which then have #ifdef HAVE_OSPRAY checks.

For instance,  avt/VisWindow/Colleagues/VisWinRendering.C

which might not have #include <visit-config.h> but use rendering atts.  I was also finding that some files via other includes would have it but not others so it was also an attempt to make sure it was there regardless.

biagas replied:
I that visit-config.h should be included where it is needed, not by transitive inclusion.
HAVE_OSPRAY isn't used in this header, so isn't needed here.

ARSanderson committed:
Merge branch 'task/biagas/investigate_vtk9_master' into task/allen/vtk9_master_ospray

ARSanderson commented:
@biagas I just resolved the conflicts.

@markcmiller86 what is the easiest way to get a list of all files that have been touched?? From that I can probably given some direction on dividing things up.

biagas commented:
@allen, I had already done the merge and resolved the conflicts. Looks like you used an outdated version of my vtk-master branch for your merge, so you have reversed some of the changes I made. You need to revert that commit please.

ARSanderson committed:
Revert "Merge branch 'task/biagas/investigate_vtk9_master' into task/allen/vtk9_master_ospray"

This reverts commit 8bce656d88f313c05ab993b7b48c8741946b90ab, reversing
changes made to 6c1d00f58eca10897c80e8f455d8576d3205dc9d.

biagas commented on src/plots/Volume/avtVolumePlot.C:
```diff
@@ -549,48 +494,104 @@ bool GetLogicalBounds(avtDataObject_p input,int &width,int &height, int &depth)
-             return true;
-         return false;
+        // The number of domains is from the database and does not
+        // relfect the actual number of doamins per rank at run time.
```
spelling: reflect, domains

biagas commented on src/plots/Volume/avtVolumePlot.C:
```diff
@@ -657,106 +657,252 @@ avtVolumePlot::ApplyRenderingTransformation(avtDataObject_p input)
+        if( atts.GetResampleType() == VolumeAttributes::NoResampling &&
+            mustResample )
+        {
+            avtCallback::IssueWarning("'No resampling' was selecting but the plot 'thinks' resampling needs to occur. This need may or may not be correct. As such, the rendered results may or may not be correct.");
```
selected instead of selecting

biagas reviewed (changes requested):
Some minor fixes eg typos.

biagas commented on src/test/tests/rendering/volume.py:
```diff
@@ -51,6 +51,7 @@
 
 AddPlot("Volume", "Primitive Var _number_0")
 vol_atts = VolumeAttributes()
+vol_atts.rendererType =  volAtts.Serial
```
volAtts should be vol_atts here.

ARSanderson committed:
removed unify cade as it is doen in the filter

ARSanderson committed:
fixed typos and whitespace

ARSanderson committed:
bug fix - need a custom range to ignore no data values

ARSanderson committed:
changed to LLNL copyright

ARSanderson committed:
fixed an issue when swaping between vtk and OSPRay

ARSanderson committed:
added defaults and checks

ARSanderson committed:
added the right condition

markcmiller86 commented:
@ARSanderson I have neglected to ask until now....what are your thoughts on the possibility of using the volume rendering infrastructure we have here now to render Constructive Solid Geometry meshes. Do you think anything here might help to facilitate that? If not, do you have suggestions for where one might try to integrate CSG mesh rendering with our VR infrastructure? Reason I ask is that I think ray-tracing/casting algorithms can do a much better job of CGS rendering than surface-based algorithms.

markcmiller86 commented:
Note that at least for the CSG stuff from Silo, every solid is a quadric function. So, its easy to use that to define an analytic expression for an opacity transition. But, not that I think about it, this is all re-sampling onto a uniform grid to do the ray-casting anyways, right? Ok, forget I asked 😞

ARSanderson committed:
removed accidently checked in file

ARSanderson committed:
added define for dumping intermediate images

ARSanderson committed:
additional debuging

ARSanderson committed:
whitespace and comment cleanup

ARSanderson committed:
added additional tests

ARSanderson committed:
updated test

ARSanderson committed:
added the ospray test back in

biagas commented:
@ARSanderson do you feel this is in good enough state for final review and merge?

ARSanderson commented:
At this point I feel it is ready to be merged. I am sure additional eyes would find minor nits. But it is probably time to merge it and start having others use it.

ARSanderson commented:
> @ARSanderson I have neglected to ask until now....what are your thoughts on the possibility of using the volume rendering infrastructure we have here now to render Constructive Solid Geometry meshes.

@markcmiller86 - I have asked Dave DeMarle to chime in as he can probably better answer than me.

demarle commented:
Hey guys! I agree that direct CSG rendering via RayTracing is likely a productive development path after this.

The shortest path I can think of is to expose a GUI in visit into OSPRay/OpenVKL's implicit isofurfaces and then map the CSG geometry onto a field defined by the quadric functions that Mark is talking about. A more experimental path would be a custom geometry module/plugin for CSG geometry implemented at the OSPRay level and then packaged with and exposed in VisIt.

Interesting historical note, the first interactive ray tracer back in late 80's was rendering CSG for the DoD.

biagas commented on src/avt/Filters/avtVisItVTKDeviceBase.h:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Lawrence Livermore National Security, LLC and other VisIt
```
Is this file still needed? It doesn't seem to be used anywhere.

biagas commented on src/avt/View/avtViewInfo.C:
```diff
@@ -293,20 +295,29 @@ avtViewInfo::SetCameraFromView(vtkCamera *vtkcam) const
+    // used in the vtkOSPRayCameraNode so instead use the Zoom here and
+    // in the Navigate3D.C and Zoom3D.C pan the camera rather than the
+    // image.
+#ifdef HAVE_OSPRAY
```
Question:  Just because we HAVE_OSPRAY doesn't mean we are currently USING OSPRAY, so doesn't this change normal operation inadvertently?

And yes, I see the logic is similar to previous, but I think the question is still valid.

biagas commented on src/avt/VisWindow/Interactors/Navigate3D.C:
```diff
@@ -91,8 +91,14 @@ Navigate3D::OnTimer(void)
+#else
         PanImage3D(Pos[0], Pos[1]);
-
+#endif
```
If HAVE_OSPRAY is true, the pan behavior is being changed whether we are using OSPRAY or not.  This doesn't seem right.

biagas commented on src/avt/VisWindow/Interactors/Zoom3D.C:
```diff
@@ -80,8 +80,14 @@ Zoom3D::OnTimer(void)
+          // Currently the SetWindowCenter called from avtViewInfo.C
+          // does not get used in the vtkOSPRayCamerNode so instead
+          // pan the camera rather than the image.
+#ifdef HAVE_OSPRAY
```
See comment for Navigate3D.

biagas commented:
Here's a link to the generated docs for the changes to [Volume Plot](https://visit-sphinx-github-user-manual.readthedocs.io/en/task-allen-vtk9_master_ospray/gui_manual/Plots/PlotTypes/VolumePlot.html)

biagas commented:
@brugger, @cyrush, @markcmiller86   Here is a list of files that could use more eyes than mine in review:

plots/Volume/QvisVolumePlotWindow
plots/Volume/avtICETCompositor.C
plots/Volume/avtLowerResolutionVolumeFilter.C
plots/Volume/avtVisItVTKRenderFilter.C
plots/Volume/avtVisItVTKRenderer.C
plots/Volume/avtVolumeFilter.C
plots/Volume/avtVolumePlot.C
plots/Volume/avtVolumeRenderer.C
plots/Volume/VolumeAttributes.code
plots/Volume/Volume.xml
avt/Filters/avtResampleFilter
avt/Pipeline/Data/avtOpacityMap
gui/QvisRenderingWindow
avt/VisWindow/Colleagues/VisWinRendering

biagas committed:
HAVE_OSPRAY usage updates

Added 'useOSPRay' to View and Interactors to test if ospray-code-paths should be taken. This will be TRUE only if HAVE_OSPRAY is defined and ospray is currently being used.
Added inclusion of visit-config.h in source and headers where HAVE_OSPRAY is used.
Whitespace cleanup in files touched.

biagas committed:
Delete unused file.

ARSanderson commented:
@biagas - THANKS Kathleen for ALL your help and additions. Greatly appreciated. I will let you pull the trigger on the merge.
